### PR TITLE
`pkg_resources`: Updates from upstream typing merge

### PR DESCRIPTION
**Pull request by Avasam**
So why even complete `pkg_resources` in the first place, if it's considered deprecated in favor of `importlib.resources`, `importlib.metadata` and their backports?
Because I've started merging the stubs directly into the setuptools repo, with the goal of being able to take it out of typeshed. And in doing so, found a lot of improvements to the typeshed stubs. At that point I was a few changes away from completing it, so here it is.

This also allowed me to cross-validate a handful of annotations, and I'll use the typeshed stubs as the reference for an upcoming PR that finishes typing `pkg_resources` upstream.

### CODE REVIEW

File: stubs/setuptools/pkg_resources/__init__.pyi [hunk 1]
```diff
@@ -1,72 +1,87 @@
...
 from re import Pattern
-from typing import IO, Any, ClassVar, Literal, Protocol, TypeVar, overload, type_check_only
+from typing import IO, Any, BinaryIO, ClassVar, Final, Literal, NamedTuple, NoReturn, Protocol, TypeVar, overload, type_check_only
 from typing_extensions import Self, TypeAlias
+from zipfile import ZipInfo
+from zipimport import zipimporter
 
 from ._vendored_packaging import requirements as packaging_requirements, version as packaging_version
 
+# using `map[Requirement]` directly causes errors with pytype https://github.com/google/pytype/issues/1324
```
Comment (JelleZijlstra) [inline stubs/setuptools/pkg_resources/__init__.pyi:16]: Even if pytype allowed it I'm not sure using the `map` type directly would be a good idea.

**Comment by Avasam** [comment]
Self-reviewing the amount of changes in 1 file, I'll put this back as draft and split it up in smaller chunks.

File: stubs/setuptools/pkg_resources/__init__.pyi [hunk 2]
```diff
@@ -316,43 +331,47 @@ class NullProvider:
...
-    def get_resource_string(self, manager: ResourceManager, resource_name): ...
-    def has_resource(self, resource_name) -> bool: ...
-    def has_metadata(self, name: str) -> bool | None: ...
+    def get_resource_filename(self, manager: ResourceManager, resource_name: str) -> str: ...
+    def get_resource_stream(self, manager: ResourceManager, resource_name: str) -> BytesIO: ...
+    def get_resource_string(self, manager: ResourceManager, resource_name: str) -> bytes: ...
+    def has_resource(self, resource_name: str) -> bool: ...
+    # Note: runtime forgets to coerce the return type to boolean,
+    # but we need to specify bool for NullProvider to respect the IResourceProvider protocol
+    # TODO: raise and fix in setuptools' repo
```
Comment (Avasam) [inline stubs/setuptools/pkg_resources/__init__.pyi:340]:
```suggestion
    # https://github.com/pypa/setuptools/pull/4254
```

File: stubs/setuptools/pkg_resources/__init__.pyi [hunk 3]
```diff
@@ -316,43 +331,47 @@ class NullProvider:
...
+    def has_metadata(self, name: str) -> bool: ...
     def get_metadata(self, name: str) -> str: ...
-    def get_metadata_lines(self, name: str) -> Generator[str, None, None]: ...
-    def resource_isdir(self, resource_name) -> bool: ...
+    def get_metadata_lines(self, name: str) -> chain[str]: ...
+    def resource_isdir(self, resource_name: str) -> bool: ...
+    # Note: runtime forgets to coerce the return type to boolean,
+    # but we need to specify bool for NullProvider to respect the IResourceProvider protocol
+    # TODO: raise and fix in setuptools' repo
```
Comment (Avasam) [inline stubs/setuptools/pkg_resources/__init__.pyi:347]:
```suggestion
    # https://github.com/pypa/setuptools/pull/4254
```

File: stubs/setuptools/pkg_resources/__init__.pyi [hunk 4]
```diff
@@ -396,6 +395,8 @@ class DistInfoDistribution(Distribution):
     EQEQ: ClassVar[Pattern[str]]
 
 class EggProvider(NullProvider):
+    egg_name: str
+    egg_info: str
```
Comment (Avasam) [inline stubs/setuptools/pkg_resources/__init__.pyi:399]:
```suggestion
    egg_name: str
    egg_info: str
```
Actually no assurance this isn't `None`, even if it feels like it should

File: stubs/setuptools/pkg_resources/__init__.pyi [hunk 5]
```diff
@@ -260,11 +259,11 @@ def get_provider(moduleOrReq: str) -> IResourceProvider: ...
 def get_provider(moduleOrReq: Requirement) -> Distribution: ...
 
 class IMetadataProvider(Protocol):
-    def has_metadata(self, name: str) -> bool | None: ...
+    def has_metadata(self, name: str) -> bool: ...
```
Comment (Avasam) [inline stubs/setuptools/pkg_resources/__init__.pyi:262]: https://github.com/pypa/setuptools/pull/4254

**Pull request merged**